File: src/test/java/io/github/vpavic/traintracker/infrastructure/fetcher/hr/HrVoyageFetcherTests.java
package io.github.vpavic.traintracker.infrastructure.fetcher.hr;

import java.net.URL;
import java.net.http.HttpClient;
import java.net.http.HttpResponse;
import java.nio.charset.Charset;
import java.nio.file.Files;
import java.nio.file.Path;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import io.github.vpavic.traintracker.domain.model.voyage.Voyage;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.BDDMockito.given;

class HrVoyageFetcherTests {

    private AutoCloseable mocks;

    @Mock
    private HttpClient httpClient;

    @Mock
    private HttpResponse<String> httpResponse;

    private HrVoyageFetcher voyageFetcher;

    @BeforeEach
    void setUp() {
        this.mocks = MockitoAnnotations.openMocks(this);
        this.voyageFetcher = new HrVoyageFetcher(this.httpClient);
    }

    @AfterEach
    void tearDown() throws Exception {
        this.mocks.close();
    }

    @Test
    void getCountry_ShouldReturnHr() {
        assertThat(this.voyageFetcher.getCountry()).isEqualTo("hr");
    }

    @Test
    void getVoyage_VoyageDoesNotExist_ShouldReturnNull() throws Exception {
        given((this.httpResponse.body())).willReturn("<html/>");
        given(this.httpClient.<String>send(any(), any())).willReturn(this.httpResponse);
        assertThat(this.voyageFetcher.getVoyage("123")).isNull();
    }

    @Test
    void getVoyage_VoyageExists_ShouldReturnVoyage() throws Exception {
        URL resource = getClass().getClassLoader().getResource("hr-tpvl-ok.html");
        assertThat(resource).isNotNull();
        String responseBody = Files.readString(Path.of(resource.toURI()), Charset.forName("Cp1250"));
        given(this.httpResponse.body()).willReturn(responseBody);
        given(this.httpClient.<String>send(any(), any())).willReturn(this.httpResponse);
        Voyage voyage = this.voyageFetcher.getVoyage("211");
        assertThat(voyage).isNotNull();
        assertThat(voyage.getCarrier().getId()).isEqualTo("hr");
    }

}

File: src/test/kotlin/io/github/vpavic/traintracker/infrastructure/fetcher/hr/HrVoyageFetcherTests.kt
/*
 * Copyright 2021 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.github.vpavic.traintracker.infrastructure.fetcher.hr

import io.mockk.MockKAnnotations
import io.mockk.every
import io.mockk.impl.annotations.InjectMockKs
import io.mockk.impl.annotations.MockK
import io.mockk.mockk
import org.apache.http.client.methods.CloseableHttpResponse
import org.apache.http.entity.FileEntity
import org.apache.http.entity.StringEntity
import org.apache.http.impl.client.CloseableHttpClient
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.io.File

class HrVoyageFetcherTests {

    @MockK
    private lateinit var httpClient: CloseableHttpClient

    @InjectMockKs
    private lateinit var voyageFetcher: HrVoyageFetcher

    @BeforeEach
    fun setUp() {
        MockKAnnotations.init(this)
    }

    @Test
    fun `Assert that voyage fetcher country is Croatia`() {
        assertThat(voyageFetcher.country).isEqualTo("hr")
    }

    @Test
    fun `Assert that fetching an unknown train returns null`() {
        val response = mockk<CloseableHttpResponse>()
        every { response.entity } returns StringEntity("<html/>")
        every { httpClient.execute(any()) } returns response
        assertThat(voyageFetcher.getVoyage("123")).isNull()
    }

    @Test
    fun `Assert that fetching a valid train returns voyage`() {
        val response = mockk<CloseableHttpResponse>()
        every { response.entity } returns FileEntity(File(javaClass.getResource("/hr-tpvl-ok.html").toURI()))
        every { httpClient.execute(any()) } returns response
        val voyage = voyageFetcher.getVoyage("211")
        assertThat(voyage).isNotNull
        assertThat(voyage?.carrier?.id).isEqualTo("hr")
    }
}
